### PR TITLE
feat(nuxt): update to `vue-bundle-renderer@0.4`

### PR DESCRIPTION
**Pull request by danielroe**
### 🔗 Linked issue

resolves https://github.com/nuxt/framework/issues/4475, resolves https://github.com/nuxt/framework/issues/4552, resolves https://github.com/nuxt/framework/issues/5508

### ❓ Type of change

- [ ] 📖 Documentation (updates to the documentation or readme)
- [x] 🐞 Bug fix (a non-breaking change that fixes an issue)
- [x] 👌 Enhancement (improving an existing functionality like performance)
- [ ] ✨ New feature (a non-breaking change that adds functionality)
- [ ] ⚠️ Breaking change (fix or feature that would cause existing functionality to change)

### 📚 Description

This PR migrates fully to the vite manifest format, normalizing the webpack manifest at build time rather than runtime.

It also adds a nitro hook (`nuxt:render:options`) which allows to customise `shouldPrefetch`, `shouldPreload` and `manifest`.

Finally, we run the actual renderer for SPA mode (but once only, with the result then reused for future requests), which avoids us having to rewrite renderer code within nuxt.

The PR is in draft to give visibility to the corresponding changes in the framework that will match https://github.com/nuxt-contrib/vue-bundle-renderer/pull/30. However, tests will fail until we can actually upgrade to a released version in this branch.

**Note**: I also move `@vite/client` to `@vite/client.mjs` so the bundle renderer can load it as a module. An alternative would be _defaulting_ to module behaviour for JS, which might be a better approach, but that will need to be implemented in https://github.com/nuxt-contrib/vue-bundle-renderer/pull/30.

### 📝 Checklist

- [x] I have linked an issue or discussion.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: packages/nuxt/src/core/runtime/nitro/renderer.ts [hunk 1]
```diff
@@ -69,7 +74,20 @@ const getSSRRenderer = lazyCachedFunction(async () => {
...
+  const manifest = await getClientManifest()
+
+  // Allow customizing options for renderer
+  const nitroApp = useNitroApp()
+  const options = {
+    manifest,
+    renderToString: () => '<div id="__nuxt"></div>',
+    buildAssetsURL
+  }
+  nitroApp.hooks.callHook('nuxt:render:options', options)
```
Comment (pi0) [inline packages/nuxt/src/core/runtime/nitro/renderer.ts:86]: (general note: We should await on `callHook`. Hookable hooks are awaitable by default)

Also what are possible usecases are you thinking to use this hook? I think we can reiterate and think better about customization interface.

File: packages/vite/src/manifest.ts [hunk 2]
```diff
@@ -1,27 +1,32 @@
...
-    async: [],
-    modules: {}
+  const devClientManifest: Manifest = {
+    '@vite/client': {
+      isEntry: true,
+      file: '@vite/client',
+      css,
+      // placeholders for future use
+      assets: [],
+      dynamicImports: [],
```
Comment (pi0) [inline packages/vite/src/manifest.ts:20]: Can normalizer add empty fields?

File: packages/nuxt/src/core/runtime/nitro/renderer.ts (hunk 1 above)
Comment (danielroe) [reply packages/nuxt/src/core/runtime/nitro/renderer.ts:86]: This is about users providing `shouldPreload` and `shouldPrefetch`.

**Comment by danielroe** [comment]
TODO items from my perspective:

- [ ] decide on whether we keep `nuxt:render:options` in this PR
- [ ] release `vue-bundle-renderer` and update resolution in this PR

Then I think it's ready to leave draft.

File: packages/nuxt/src/core/runtime/nitro/renderer.ts (hunk 1 above)
Comment (pi0) [reply packages/nuxt/src/core/runtime/nitro/renderer.ts]: Sure. We can add it back. I need to try new renderer more. We might find sexier interface than vue2 shouldPreload/prefetch callbacks.

**Comment by pi0** [comment]
- Released `vue-bundle-renderer@0.4.0`
- You can remove customization hook for this PR 👍🏼

**Comment by dungonfq** [comment]
Awesome work @danielroe @pi0 🙌🏻 🎉 Thank u a lots for these PRs, we'd waited for it for a long time.

When will we have documentation about how to customize `shouldPrefetch`, `shouldPreload` and `manifest`?

**Comment by danielroe** [comment]
@dungonfq The ability customise the renderer options was [removed shortly before merge](https://github.com/nuxt/framework/pull/6210#discussion_r937950236) for faster iteration.

I will update the description.

**Comment by dungonfq** [comment]
@danielroe could u give me where & how to use these new options `shouldPrefetch`,` shouldPreload` and `manifest`?

At first, I thought I can config them in [nuxt.config.ts](https://v3.nuxtjs.org/api/configuration/nuxt.config).

**Comment by danielroe** [comment]
It is not yet possible.

**Comment by dungonfq** [comment]
Thank @danielroe , my last confusion:

So, after u merge this PR. The prefetch issue will be fixed automatically, right?
https://github.com/nuxt/framework/issues/4475

What I expected is all my modules weren't prefetched unnecessarily. Then when I load a page, only the required components/modules, which the page is using, will be fetched.

**Comment by danielroe** [comment]
@dungonfq I don't know what you are asking, but if you are experiencing a new issue, please check that it is not resolved on the edge channel, and then open a new issue and I'll look into it 🙏

**Comment by kosmeln** [comment]
Hello @danielroe . 
Would you help me to understand if this ticket resoled https://github.com/nuxt/framework/issues/4475 issue? I'm still experiencing it after upgrading to RC-7, but was thinking if this is something you mentioned that was branched of for the next release.

**Comment by danielroe** [comment]
@kosmeln Yes, it should have resolved it. (But see https://github.com/nuxt/framework/issues/4475#issuecomment-1162818358.)

If you are experiencing a bug, please raise a new issue with a reproduction. :pray:

**Pull request merged**